src/main/java/org/kuali/kfs/module/ar/businessobject/TimeBasedBillingPeriod.java

```diff
@@ -0,0 +1,132 @@
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.kuali.kfs.module.ar.businessobject;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.joda.time.DateTimeZone;
+import org.joda.time.LocalDate;
+import org.kuali.kfs.coa.businessobject.AccountingPeriod;
+import org.kuali.kfs.coa.service.AccountingPeriodService;
+import org.kuali.kfs.module.ar.ArConstants;
+
+import java.sql.Date;
+
+public class TimeBasedBillingPeriod extends BillingPeriod {
+    
+    /*
+     * CU Customization
+     */
+    private static final Logger LOG = LogManager.getLogger(TimeBasedBillingPeriod.class);
+    
+    public TimeBasedBillingPeriod(ArConstants.BillingFrequencyValues billingFrequency, Date awardStartDate, Date currentDate, Date lastBilledDate, AccountingPeriodService accountingPeriodService) {
+        super(billingFrequency, awardStartDate, currentDate, lastBilledDate, accountingPeriodService);
+    }
+
+    @Override
+    protected Date determineEndDateByFrequency() {
+        final AccountingPeriod accountingPeriod = findPreviousAccountingPeriod(currentDate);
+        return accountingPeriod.getUniversityFiscalPeriodEndDate();
+    }
+
+    protected Integer calculatePreviousPeriodByFrequency(Integer currentAccountingPeriodCode, int periodsInBillingFrequency) {
+        Integer previousAccountingPeriodCode;
+        final int subAmt = (currentAccountingPeriodCode % periodsInBillingFrequency) == 0 ? periodsInBillingFrequency : currentAccountingPeriodCode % periodsInBillingFrequency;
+
+        previousAccountingPeriodCode = currentAccountingPeriodCode - subAmt;
+        return previousAccountingPeriodCode;
+    }
+
+    @Override
+    protected boolean canThisBeBilledByBillingFrequency() {
+        if (ArConstants.BillingFrequencyValues.ANNUALLY.equals(billingFrequency) && accountingPeriodService.getByDate(lastBilledDate).getUniversityFiscalYear() >= accountingPeriodService.getByDate(currentDate).getUniversityFiscalYear()) {
+            return false;
+        } else if (StringUtils.equals(findPreviousAccountingPeriod(currentDate).getUniversityFiscalPeriodCode(), findPreviousAccountingPeriod(lastBilledDate).getUniversityFiscalPeriodCode()) &&
+            accountingPeriodService.getByDate(lastBilledDate).getUniversityFiscalYear().equals(accountingPeriodService.getByDate(currentDate).getUniversityFiscalYear())) {
+            return false;
+        }
+
+        return true;
+    }
+    
+    /*
+     * CU Customization
+     */
+    @Override
+    protected Date determineStartDateByFrequency() {
```

Comment:
Once concern that I have is whether this method is only used for the determining the next billing period start date. I want to make sure it does not inadvertently affect offer parts of the application.

Reply:
Currently this function is used by by ContractsAndGrantsInvoiceCreateDocumentService and by VerifyBillingFrequencyServiceImpl.  The VerifyBillingFrequencyServiceImpl functions take in ContractsAndGrantsBillingAward objects as attributes.  I think this object is restricted to CG Invoice & Award processing.  I think changes here will be restricted to the CG module.  I am not sure if there is something we should do so we notice if something changes about that.

Reply:
thank you for clarifying this